src/exp.ml

```diff
@@ -887,7 +887,10 @@ and import_let_fun
     | _ ->
       raise None Unexpected "A variable name instead of a pattern was expected"
     ) >>= fun x ->
+    let predefined_variables = List.map snd (Name.Map.bindings typ_vars) in
     Type.of_typ_expr true typ_vars vb_expr.exp_type >>= fun (e_typ, typ_vars, new_typ_vars) ->
+    let* e_typ = Type.decode_var_tags new_typ_vars false e_typ in
+    let new_typ_vars = VarEnv.remove_many predefined_variables new_typ_vars in
```

Comment:
Basically make sure to save what are the variables seen before and remove it later in the function header

Reply:
As a comment in the code? (and the same for all the next remarks)